File: lib/services/recent_projects_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/recent_project.dart';

class RecentProjectsService extends ChangeNotifier {
  static const String _recentProjectsKey = 'recent_projects';
  static const int _maxRecentProjects = 10;

  List<RecentProject> _recentProjects = [];
  bool _isLoaded = false;

  List<RecentProject> get recentProjects => _recentProjects;
  bool get isLoaded => _isLoaded;

  /// Load recent projects from SharedPreferences
  Future<void> loadRecentProjects() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final jsonString = prefs.getString(_recentProjectsKey);

      if (jsonString != null) {
        final List<dynamic> jsonList = json.decode(jsonString);
        _recentProjects = jsonList
            .map((json) => RecentProject.fromJson(json))
            .toList()
          ..sort((a, b) => b.lastOpened.compareTo(a.lastOpened));
      }

      _isLoaded = true;
      notifyListeners();
    } catch (e) {
      print('Error loading recent projects: $e');
      _recentProjects = [];
      _isLoaded = true;
      notifyListeners();
    }
  }

  /// Add or update a project in the recent list
  Future<void> addRecentProject({
    required String name,
    required String path,
  }) async {
    try {
      // Remove if already exists
      _recentProjects.removeWhere((p) => p.path == path);

      // Add at the beginning with current time
      final newProject = RecentProject(
        name: name,
        path: path,
        lastOpened: DateTime.now(),
      );
      _recentProjects.insert(0, newProject);

      // Limit to max recent projects
      if (_recentProjects.length > _maxRecentProjects) {
        _recentProjects = _recentProjects.take(_maxRecentProjects).toList();
      }

      await _saveRecentProjects();
      notifyListeners();
    } catch (e) {
      print('Error adding recent project: $e');
    }
  }

  /// Remove a project from the recent list
  Future<void> removeRecentProject(String path) async {
    try {
      _recentProjects.removeWhere((p) => p.path == path);
      await _saveRecentProjects();
      notifyListeners();
    } catch (e) {
      print('Error removing recent project: $e');
    }
  }

  /// Clear all recent projects
  Future<void> clearRecentProjects() async {
    try {
      _recentProjects.clear();
      await _saveRecentProjects();
      notifyListeners();
    } catch (e) {
      print('Error clearing recent projects: $e');
    }
  }

  /// Save recent projects to SharedPreferences
  Future<void> _saveRecentProjects() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final jsonList = _recentProjects.map((p) => p.toJson()).toList();
      final jsonString = json.encode(jsonList);
      await prefs.setString(_recentProjectsKey, jsonString);
    } catch (e) {
      print('Error saving recent projects: $e');
    }
  }

  /// Check if a project path still exists (platform-specific)
  /// For now, we'll keep all projects and let the user handle invalid paths
  /// In the future, could add file system checks here
  Future<bool> projectExists(String path) async {
    // TODO: Implement platform-specific existence checks
    // For now, assume all projects exist
    return true;
  }
}

File: react-native-writr/src/services/RecentProjectsService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import {RecentProject} from '../types';
import {RecentProjectFactory} from '../models/RecentProject';

const RECENT_PROJECTS_KEY = '@writr_recent_projects';
const MAX_RECENT_PROJECTS = 10;

/**
 * Service for managing recent projects
 * Uses AsyncStorage for persistence
 */
export class RecentProjectsService {
  /**
   * Load all recent projects
   */
  async loadRecentProjects(): Promise<RecentProject[]> {
    try {
      const data = await AsyncStorage.getItem(RECENT_PROJECTS_KEY);
      if (!data) {
        return [];
      }

      const projects: RecentProject[] = JSON.parse(data);
      return RecentProjectFactory.sortByLastOpened(projects);
    } catch (error) {
      console.error('Error loading recent projects:', error);
      return [];
    }
  }

  /**
   * Add or update a recent project
   */
  async addRecentProject(name: string, path: string): Promise<void> {
    try {
      const projects = await this.loadRecentProjects();

      // Remove existing entry if present
      const filtered = projects.filter((p) => p.path !== path);

      // Add new entry at the beginning
      const newProject = RecentProjectFactory.create(name, path);
      filtered.unshift(newProject);

      // Limit to MAX_RECENT_PROJECTS
      const limited = filtered.slice(0, MAX_RECENT_PROJECTS);

      await AsyncStorage.setItem(
        RECENT_PROJECTS_KEY,
        JSON.stringify(limited),
      );
    } catch (error) {
      console.error('Error adding recent project:', error);
    }
  }

  /**
   * Remove a project from recent list
   */
  async removeRecentProject(path: string): Promise<void> {
    try {
      const projects = await this.loadRecentProjects();
      const filtered = projects.filter((p) => p.path !== path);

      await AsyncStorage.setItem(
        RECENT_PROJECTS_KEY,
        JSON.stringify(filtered),
      );
    } catch (error) {
      console.error('Error removing recent project:', error);
    }
  }

  /**
   * Clear all recent projects
   */
  async clearRecentProjects(): Promise<void> {
    try {
      await AsyncStorage.removeItem(RECENT_PROJECTS_KEY);
    } catch (error) {
      console.error('Error clearing recent projects:', error);
    }
  }
}

// Export a singleton instance
export default new RecentProjectsService();
